core/src/main/java/org/jboss/sbomer/core/features/sbom/utils/GenericPurlWrapperUtil.java

```diff
@@ -0,0 +1,386 @@
+        private final String from;
+        private final String to;
+
+        public ValueChanges(String from, String to) {
+            this.from = from;
+            this.to = to;
+        }
+
+        public String getFrom() {
+            return from;
+        }
+
+        public String getTo() {
+            return to;
+        }
+    }
+
+    @Getter
+    private final PackageURL packageURL;
+
+    @Getter
+    private final double confidenceScore;
+
+    private static final ObjectMapper objectMapper = new ObjectMapper();
+    /*
+     * Our Regex that covers most the patterns we see on the RCM release area
+     */
+    private static final Pattern[] FILENAME_VERSION_REGEX_STRATEGIES = {
+            Pattern.compile("(?<version>\\d+\\.\\d+\\.\\d+)(?<qualsep>[.-](?<qualifier>Final|[A-Z]+\\d*))?"),
+            Pattern.compile("(?<version>\\d+\\.\\d+)(?<qualsep>[.-](?<qualifier>[A-Z]+\\d*))?"),
+            Pattern.compile("(?<version>\\d+_\\d+)(?<qualsep>[.-](?<qualifier>[A-Z]+\\d*))?") };
+
+    /*
+     * This is our list of PURLs we think are good examples to skew to. These are used only in the calculation of the
+     * confidence score
+     */
+    private static final List<String> IDEAL_PURLS = Arrays.asList(
+            "pkg:generic/foo-1.2.3.zip",
+            "pkg:generic/foo-3.4.5-CR2.zip",
+            "pkg:generic/foo-5.6.7.tar.gz",
+            "pkg:generic/foo-bar-7.2.4-maven-repository.zip",
+            "pkg:generic/quite-long-productname-1.2.3.DR1-maven-repository.zip",
+            "pkg:generic/somepro-duct-7.4.11-runtime-maven-repository.zip");
+
+    // We only want to run this once
+    static {
+        setupNorms();
+    }
+
+    // If we just want to use the class static
+    public GenericPurlWrapperUtil() {
+        packageURL = null;
+        confidenceScore = 0.0;
+    }
+
+    public GenericPurlWrapperUtil(String purl) throws MalformedPackageURLException {
+        this(new PackageURL(purl));
+    }
+
+    public GenericPurlWrapperUtil(PackageURL purl) throws MalformedPackageURLException {
+        packageURL = purl;
+        if (!packageURL.getType().equals(PackageURL.StandardTypes.GENERIC))
+            throw new MalformedPackageURLException("Not a Generic type PURL");
+        if (packageURL.getVersion() != null)
+            throw new MalformedPackageURLException("PURL already has version defined");
+        confidenceScore = genericPurlVersionConfidence(packageURL.canonicalize());
+    }
+
+    /*
+     * Values of penalties we can tune
+     */
+    private static final double LENGTH_WEIGHT = 0.005;
+    private static final double DIGIT_RATIO_WEIGHT = 5.0;
+    private static final double LETTER_RATIO_WEIGHT = 1.5;
+    private static final double SEPARATOR_RATIO_WEIGHT = 4.0;
+    private static final double SEPARATOR_TYPES_WEIGHT = 0.1;
+    private static final double MAX_DIGIT_RUN_WEIGHT = 0.2;
+    private static final double INTERSTITIAL_LETTER_WEIGHT = 0.02;
```

Comment:
TBH, I have no idea if this is good or bad, but I see all the test cases pass. It seems the deliverables don't follow a specific specification for the version, so that doesn't help.